flink-table/flink-table-storage/src/test/java/org/apache/flink/table/storage/filestore/SortBufferMemTableTest.java

```diff
@@ -0,0 +1,139 @@
+        table.clear();
+        innerTestRandom(Integer.MAX_VALUE);
+    }
+
+    private void innerTestRandom(int numRecords) throws IOException {
+        int bound = 100;
+        Random rnd = new Random();
+        List<KeyValue> inputs = new ArrayList<>();
+        Set<Integer> usedSequenceNumber = new HashSet<>();
+        for (int i = 0; i < numRecords; i++) {
+            int value = rnd.nextInt(bound);
+            int sequenceNumber;
+            while (true) {
+                sequenceNumber = rnd.nextInt();
+                if (!usedSequenceNumber.contains(sequenceNumber)) {
+                    usedSequenceNumber.add(sequenceNumber);
+                    break;
+                }
+            }
+            KeyValue keyValue =
+                    new KeyValue()
+                            .replace(
+                                    GenericRowData.of(value),
+                                    sequenceNumber,
+                                    rnd.nextBoolean() ? ValueKind.ADD : ValueKind.DELETE,
+                                    GenericRowData.of(value));
+            boolean success =
+                    table.put(
+                            keyValue.sequenceNumber(),
+                            keyValue.valueKind(),
+                            keyValue.key(),
+                            keyValue.value());
+            if (!success) {
+                if (numRecords == Integer.MAX_VALUE) {
+                    break;
+                } else {
+                    throw new EOFException();
+                }
+            }
+            inputs.add(keyValue);
+        }
+
+        Assert.assertEquals(inputs.size(), table.size());
+
+        RecordReader<KeyValue> reader = table.createOrderedReader();
+
+        RecordIterator<KeyValue> iterator = reader.readBatch();
+        List<KeyValue> results = new ArrayList<>();
+        Assert.assertNotNull(iterator);
+        while (iterator.advanceNext()) {
+            KeyValue current = iterator.current();
+            if (results.size() > 0) {
+                Assert.assertTrue(isEquals(results.get(results.size() - 1), iterator.previous()));
+            }
+            results.add(current.copy(INT_SERIALIZER, INT_SERIALIZER));
+        }
+        Assert.assertNull(reader.readBatch());
+
+        inputs.sort(
+                Comparator.<KeyValue>comparingInt(o -> o.key().getInt(0))
+                        .thenComparingLong(KeyValue::sequenceNumber));
+
+        Assert.assertEquals(results.size(), inputs.size());
+        for (int i = 0; i < results.size(); i++) {
+            Assert.assertTrue(isEquals(inputs.get(i), results.get(i)));
+        }
+    }
+
+    private boolean isEquals(KeyValue kv1, KeyValue kv2) {
+        return kv1.sequenceNumber() == kv2.sequenceNumber()
+                && kv1.valueKind() == kv2.valueKind()
+                && kv1.key().getInt(0) == kv2.key().getInt(0)
+                && kv1.value().getInt(0) == kv2.value().getInt(0);
```

Comment:
Why not add a `hashCode` and `equals` method to `KeyValue`. Other tests can also use them.

Reply:
For this case, the expected row is `GenericRowData`, but the row in returned `KeyValue` is `OffsetRowData`. They can not be compared.